### DISCUSSION TITLE
CORS Policy

### DISCUSSION THREAD

**Discussion by stefanobarcellos** [mercadopago/sdk-js#213] [Q&A] [answered]
Na hora de fazer uma assinatura via javascript utilizando essa SDK, eu recebo um block em relação ao meu domínio.

O endereço, por exemplo:

fetch("https://api.mercadopago.com/preapproval", {
method: "POST"
            
bloqueia por causa da origem.

Lá na aplicação eu coloquei meu domínio corretamente, mas ele continua não aceitando a solicitação. A aplicação está configurada para assinaturas. Estou utilizando cartões e credenciais de teste corretamente.

O erro no console do navegador:

**Access to fetch at 'https://api.mercadopago.com/preapproval' from origin 'https://meusite.com.br' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status.**

Onde devo adicionar meu domínio para que o servidor do mercado pago não bloqueie minha solicitação desse domínio? Ou alguma outra correção?

Obrigado.

==================================

ENGLISH:

When I make a subscription via javascript using this SDK, I get a block regarding my domain.

The address, for example:

fetch("https://api.mercadopago.com/preapproval", {
method: "POST"

is blocked because of the origin.

In the application I entered my domain correctly, but it still does not accept the request. The application is configured for subscriptions. I am using cards and test credentials correctly.

The error in the browser console:

**Access to fetch at 'https://api.mercadopago.com/preapproval' from origin 'https://mysite.com.br' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status.**

Where should I add my domain so that the Mercado Pago server does not block my request for this domain? Or any other fix?

Thank you.

**Comment by icaldana** [comment]
Olá @stefanobarcellos 

Esse tipo de problema geralmente está relacionado com a implementação do seu backend, pois do ponto de vista da SDK JS não há nenhum controle que restrinja esse ponto. Você está utilizando alguma [SDK de backend](https://www.mercadopago.com.br/developers/pt/docs/sdks-library/server-side)? Em caso positivo, você poderia perguntar diretamente na aba de discussions da SDK em questão. 

Outra opção seria criar um ticket de suporte em https://www.mercadopago.com/developers/pt/support/center/contact.

**Comment by stefanobarcellos** [comment]
Agora estou utilizando a mercadopago/sdk-js apenas para pegar o token do cartão. Que passa para o backend em PHP que faz a assinatura por meio de um requisição CURL em PHP. Sem utilizar a SDK em PHP do Mercado Pago. Assim, resolvi o erro de CORS Policy.

A dúvida agora é como integrar o pagamento via transferência bancária para também aceitar PIX.

Eu queria fazer um sistema de assinatura com pagamento de crédito, débito e PIX. Alguma dica?

**Reply by icaldana** [reply, marked as answer]
Olá! Nesse caso, te recomendaria utilizar a solução front-end de Checkout Bricks, no caso o [Payment Brick](https://www.mercadopago.com.br/developers/pt/docs/checkout-bricks/payment-brick/introduction) para coletar os dados do cliente e gerar o token do cartão, porém, como você já implementou essa parte eu deixo a seu critério.

Sobre o backend, você poderia utilizar a [SDK de PHP](https://github.com/mercadopago/sdk-php/tree/master/src/MercadoPago/Resources/PreApproval) para criar a assinatura, ou também fazer direto [via API](https://www.mercadopago.com.br/developers/pt/reference/subscriptions/_preapproval/post).

Por último, gostaria de comentar que pagamentos recorrentes geralmente são feitos no crédito, o uso de PIX nesse caso não seria tão recomendado.

Torcendo para sua integração dar certo, obrigado! 😄

**Comment by stefanobarcellos** [comment]
Deu certo.

Usei o bricks no front end com algumas funções javascript custom.
Ele faz a requisição para uma página em php que está usando sdk em php do mercado pago
A resposta é tratada pelo javascript da página principal indicando a finalização do pagamento com pix ou cartão.
Webhooks envia uma notificação de payment para outra página de php que está usando a API pura via get para atualizar o banco de dados.

Obs: Achei alguns erros de digitação, sintaxe e lógica na documentação. A documentação está cruzada com o site também, tá bem chatinho isso. Mas deu certo.

**Reply by icaldana** [reply]
Olá @stefanobarcellos,

Fico feliz que tenha dado certo! Sobre os erros de digitação e lógica, ficaríamos gratos se pudesse apontá-los, assim consigo direcionar às equipes responsáveis para corrigirmos. 

Obrigado!
